Custom shell completion doesn't work, root.Before() not being called

## My urfave/cli version is

`v3.9.1-0.20260524212652-be8b79d0c8de`

## Checklist

- [X] Are you running the latest v3 release? The list of releases is [here](https://github.com/urfave/cli/releases).
- [X] Did you check the manual for your release? The v3 manual is [here](https://cli.urfave.org/v3/getting-started/)
- [X] Did you perform a search about this problem? Here's the [GitHub guide](https://help.github.com/en/github/managing-your-work-on-github/using-search-to-filter-issues-and-pull-requests) about searching.

## Dependency Management

- My project is using go modules.

## Describe the bug

Custom shell completion still doesn't work because `root.Before()` is not being called. So my config struct is nil and the function which is being called to generated the completion slice panics.

## To reproduce

```go
func main() {
	conf := &cfg.Config{}

	cmd := &cli.Command{
		Commands: []*cli.Command{
			IndexShow(conf),
		},

		Before: func(ctx context.Context, cmd *cli.Command) (context.Context, error) {
			if err := conf.Init(); err != nil {
				if len(os.Args) > 1 {
					return nil, err
				} else {
					fmt.Println(cmd.UsageText)
					return nil, nil
				}
			}

			log.Init(conf)

			return nil, nil
		},
	}
}

func IndexShow(conf *cfg.Config) *cli.Command {
	return &cli.Command{
		Name:    "show",
		Aliases: []string{"sh"},
		Usage:   "show details about an index",

		Action: func(ctx context.Context, cmd *cli.Command) error {
			index := cmd.Args().Get(0)
			if index == "" {
				return errors.New("no index specified")
			}

			return es.IndexShow(conf, cmd.Args().Get(0))
		},

		ShellComplete: func(ctx context.Context, cmd *cli.Command) {
			if cmd.NArg() > 0 {
				return
			}

			indices, err := es.IndexNames(conf)
			if err != nil {
				return
			}

			for _, index := range indices {
				fmt.Println(index)
			}
		},
	}
}
```

- [original top level cli def](https://codeberg.org/scip/esctl/src/branch/main/cmd/root.go#L103)
- [completion definition](https://codeberg.org/scip/esctl/src/branch/main/cmd/index.go#L139)

## Observed behavior

The `Before()` in the top level cli is not being called and the config struct is nil when shell complete is being executed. It then crashes:

```default
./esctl index show --generate-shell-completion
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x1a5aa41]

goroutine 1 [running]:
codeberg.org/scip/esctl/pkg/es.IndexNames(0x1bf3740?)
        /home/scip/dev/esctl/pkg/es/index.go:38 +0x21
codeberg.org/scip/esctl/cmd.Index.IndexShow.func2({0x1f38e6b?, 0x4?}, 0xc00022b110?)
        /home/scip/dev/esctl/cmd/index.go:144 +0x34
github.com/urfave/cli/v3.checkCompletions({0x21bc5d0, 0xc000266420}, 0xc00023d8c8)
        /home/scip/go/pkg/mod/github.com/urfave/cli/v3@v3.9.1-0.20260524212652-be8b79d0c8de/help.go:515 +0x24e
github.com/urfave/cli/v3.(*Command).run(0xc00023d8c8, {0x21bc5d0, 0xc000266420}, {0xc000268ed0, 0x1, 0x1})
        /home/scip/go/pkg/mod/github.com/urfave/cli/v3@v3.9.1-0.20260524212652-be8b79d0c8de/command_run.go:161 +0xb7b
github.com/urfave/cli/v3.(*Command).run(0xc00023f8c8, {0x21bc5d0, 0xc000266300}, {0xc00007dd20, 0x2, 0x2})
        /home/scip/go/pkg/mod/github.com/urfave/cli/v3@v3.9.1-0.20260524212652-be8b79d0c8de/command_run.go:292 +0x20b6
github.com/urfave/cli/v3.(*Command).run(0xc000245608, {0x21bc5d0, 0xc0000a1dd0}, {0xc000020100, 0x4, 0x4})
        /home/scip/go/pkg/mod/github.com/urfave/cli/v3@v3.9.1-0.20260524212652-be8b79d0c8de/command_run.go:292 +0x20b6
github.com/urfave/cli/v3.(*Command).Run(...)
        /home/scip/go/pkg/mod/github.com/urfave/cli/v3@v3.9.1-0.20260524212652-be8b79d0c8de/command_run.go:93
codeberg.org/scip/esctl/cmd.Main()
        /home/scip/dev/esctl/cmd/root.go:119 +0x127b
main.main()
        /home/scip/dev/esctl/main.go:26 +0x13
```

## Expected behavior

I'd expect that the top level cli is being executed.

My current workaround is to load the config directly in the `SHellComplete:` function.

## Want to fix this yourself?

We'd love to have more contributors on this project! If the fix for
this bug is easily explained and very small, feel free to create a
pull request for it.

## Run `go version` and paste its output here

```
go version go1.25.0 linux/amd64
```

## Run `go env` and paste its output here

```
AR='ar'
CC='gcc'
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_ENABLED='1'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
CXX='g++'
GCCGO='gccgo'
GO111MODULE=''
GOAMD64='v1'
GOARCH='amd64'
GOAUTH='netrc'
GOBIN=''
GOCACHE='/home/scip/.cache/go-build'
GOCACHEPROG=''
GODEBUG=''
GOENV='/home/scip/.config/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFIPS140='off'
GOFLAGS=''
GOGCCFLAGS='-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=/tmp/go-build2427390874=/tmp/go-build -gno-record-gcc-switches'
GOHOSTARCH='amd64'
GOHOSTOS='linux'
GOINSECURE=''
GOMOD='/home/scip/dev/esctl/go.mod'
GOMODCACHE='/home/scip/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='linux'
GOPATH='/home/scip/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/home/scip/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64'
GOSUMDB='sum.golang.org'
GOTELEMETRY='off'
GOTELEMETRYDIR='/home/scip/.config/go/telemetry'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/home/scip/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.0.linux-amd64/pkg/tool/linux_amd64'
GOVCS=''
GOVERSION='go1.25.0'
GOWORK=''
PKG_CONFIG='pkg-config'
```

--- a/command_run.go
+++ b/command_run.go
@@ -158,7 +158,12 @@ func (cmd *Command) run(ctx context.Context, osArgs []string) (_ context.Context
 
 	tracef("using post-parse arguments %[1]q (cmd=%[2]q)", args, cmd.Name)
 
-	if checkCompletions(ctx, cmd) {
+	if shouldRunCompletion(cmd) {
+		var beforeErr error
+		if ctx, beforeErr = runBefore(ctx, commandChain(cmd)); beforeErr != nil {
+			return ctx, beforeErr
+		}
+		runCompletion(ctx, cmd)
 		return ctx, nil
 	}
 
@@ -297,23 +302,12 @@ func (cmd *Command) run(ctx context.Context, osArgs []string) (_ context.Context
 	// perform the command action.
 	//
 	// First, resolve the chain of nested commands up to the parent.
-	var cmdChain []*Command
-	for p := cmd; p != nil; p = p.parent {
-		cmdChain = append(cmdChain, p)
-	}
-	slices.Reverse(cmdChain)
+	cmdChain := commandChain(cmd)
 
 	// Run Before actions in order.
-	for _, cmd := range cmdChain {
-		if cmd.Before == nil {
-			continue
-		}
-		if bctx, err := cmd.Before(ctx, cmd); err != nil {
-			deferErr = cmd.handleExitCoder(ctx, err)
-			return ctx, deferErr
-		} else if bctx != nil {
-			ctx = bctx
-		}
+	if ctx, err = runBefore(ctx, cmdChain); err != nil {
+		deferErr = err
+		return ctx, deferErr
 	}
 
 	// Run flag actions in order.
@@ -362,4 +356,27 @@ func (cmd *Command) run(ctx context.Context, osArgs []string) (_ context.Context
 
 	tracef("returning deferErr (cmd=%[1]q) %[2]q", cmd.Name, deferErr)
 	return ctx, deferErr
+}
+
+func commandChain(cmd *Command) []*Command {
+	var cmdChain []*Command
+	for p := cmd; p != nil; p = p.parent {
+		cmdChain = append(cmdChain, p)
+	}
+	slices.Reverse(cmdChain)
+	return cmdChain
+}
+
+func runBefore(ctx context.Context, cmdChain []*Command) (context.Context, error) {
+	for _, cmd := range cmdChain {
+		if cmd.Before == nil {
+			continue
+		}
+		if bctx, err := cmd.Before(ctx, cmd); err != nil {
+			return ctx, cmd.handleExitCoder(ctx, err)
+		} else if bctx != nil {
+			ctx = bctx
+		}
+	}
+	return ctx, nil
 }

--- a/command_test.go
+++ b/command_test.go
@@ -2098,8 +2098,9 @@ func TestCommand_OrderOfOperations(t *testing.T) {
 		r := require.New(t)
 
 		_ = cmd.Run(buildTestContext(t), []string{"command", completionFlag})
-		r.Equal(1, counts.ShellComplete)
-		r.Equal(1, counts.Total)
+		r.Equal(1, counts.Before)
+		r.Equal(2, counts.ShellComplete)
+		r.Equal(2, counts.Total)
 	})
 
 	t.Run("nil on usage error", func(t *testing.T) {

--- a/completion_test.go
+++ b/completion_test.go
@@ -393,6 +393,37 @@ func TestCompletionSubcommandCustomShellComplete(t *testing.T) {
 	r.Equal("custom-index\n", out.String())
 }
 
+func TestCompletionRunsBeforeChain(t *testing.T) {
+	type contextKey struct{}
+
+	out := &bytes.Buffer{}
+	cmd := &Command{
+		EnableShellCompletion: true,
+		Writer:                out,
+		Before: func(ctx context.Context, cmd *Command) (context.Context, error) {
+			return context.WithValue(ctx, contextKey{}, "ready"), nil
+		},
+		Commands: []*Command{
+			{
+				Name: "index",
+				Commands: []*Command{
+					{
+						Name: "show",
+						ShellComplete: func(ctx context.Context, cmd *Command) {
+							fmt.Fprintln(cmd.Root().Writer, ctx.Value(contextKey{}))
+						},
+						Action: func(ctx context.Context, cmd *Command) error { return nil },
+					},
+				},
+			},
+		},
+	}
+
+	r := require.New(t)
+	r.NoError(cmd.Run(buildTestContext(t), []string{"foo", "index", "show", completionFlag}))
+	r.Equal("ready\n", out.String())
+}
+
 func TestCompletionInvalidShell(t *testing.T) {
 	cmd := &Command{
 		EnableShellCompletion: true,

--- a/completion_test.go
+++ b/completion_test.go
@@ -3,7 +3,9 @@ package cli
 import (
 	"bytes"
 	"context"
+	"errors"
 	"fmt"
+	"io"
 	"strings"
 	"testing"
 
@@ -422,6 +424,27 @@ func TestCompletionRunsBeforeChain(t *testing.T) {
 	r := require.New(t)
 	r.NoError(cmd.Run(buildTestContext(t), []string{"foo", "index", "show", completionFlag}))
 	r.Equal("ready\n", out.String())
+}
+
+func TestCompletionReturnsBeforeError(t *testing.T) {
+	beforeErr := errors.New("load config")
+	completed := false
+
+	cmd := &Command{
+		EnableShellCompletion: true,
+		Writer:                io.Discard,
+		Before: func(ctx context.Context, cmd *Command) (context.Context, error) {
+			return nil, beforeErr
+		},
+		ShellComplete: func(ctx context.Context, cmd *Command) {
+			completed = true
+		},
+	}
+
+	err := cmd.Run(buildTestContext(t), []string{"foo", completionFlag})
+
+	require.ErrorIs(t, err, beforeErr)
+	assert.False(t, completed)
 }
 
 func TestCompletionInvalidShell(t *testing.T) {

--- a/help.go
+++ b/help.go
@@ -492,7 +492,7 @@ func checkShellCompleteFlag(c *Command, arguments []string) (bool, []string) {
 	return true, arguments[:pos]
 }
 
-func checkCompletions(ctx context.Context, cmd *Command) bool {
+func shouldRunCompletion(cmd *Command) bool {
 	tracef("checking completions on command %[1]q", cmd.Name)
 
 	if !cmd.Root().shellCompletion {
@@ -509,13 +509,14 @@ func checkCompletions(ctx context.Context, cmd *Command) bool {
 	}
 
 	tracef("no subcommand found for completion %[1]q", cmd.Name)
+	return true
+}
 
+func runCompletion(ctx context.Context, cmd *Command) {
 	if cmd.ShellComplete != nil {
 		tracef("running shell completion func for command %[1]q", cmd.Name)
 		cmd.ShellComplete(ctx, cmd)
 	}
-
-	return true
 }
 
 func subtract(a, b int) int {

